e2e_playwright/multipage_apps_v2/mpa_v2_basics_test.py

```diff
@@ -0,0 +1,297 @@
+):
+    """Test that the sidebar nav is rendered correctly."""
+    assert_snapshot(themed_app.get_by_test_id("stSidebarNav"), name="mpa-sidebar_nav")
+
+
+def test_can_switch_between_pages_by_clicking_on_sidebar_links(app: Page):
+    """Test that we can switch between pages by clicking on sidebar links."""
+    app.get_by_test_id("stSidebarNav").locator("a").nth(1).click()
+    wait_for_app_run(app)
+    expect(app.get_by_test_id("stHeading").nth(1)).to_contain_text("Page 3")
+
+
+def test_main_script_persists_across_page_changes(app: Page):
+    """Test that we can switch between pages and content from main script persists."""
+    app.get_by_test_id("stSidebarNav").locator("a").nth(1).click()
+    wait_for_app_run(app)
+    expect(app.get_by_test_id("stHeading").nth(0)).to_contain_text("Main Page")
+
+
+def test_main_script_widgets_persist_across_page_changes(app: Page):
+    """Test that we can switch between pages and widgets from main script persists."""
+    slider = app.locator('.stSlider [role="slider"]')
+    slider.click()
+    slider.press("ArrowRight")
+    wait_for_app_run(app, wait_delay=500)
+
+    app.get_by_test_id("stSidebarNav").locator("a").nth(3).click()
+    wait_for_app_run(app)
+    expect(app.get_by_test_id("stMarkdown").nth(0)).to_contain_text("x is 1")
+
+
+def test_supports_navigating_to_page_directly_via_url(page: Page, app_port: int):
+    """Test that we can navigate to a page directly via URL."""
+    page.goto(f"http://localhost:{app_port}/page_5")
+    wait_for_app_loaded(page)
+
+    expect(page.get_by_test_id("stHeading").nth(1)).to_contain_text("Page 5")
+
+
+def test_supports_navigating_to_page_directly_via_url_title(page: Page, app_port: int):
+    """Test that we can navigate to a page directly via URL. using the title."""
+    page.goto(f"http://localhost:{app_port}/Different_Title")
+    wait_for_app_loaded(page)
+
+    expect(page.get_by_test_id("stHeading").nth(1)).to_contain_text("Page 3")
+
+
+def test_can_switch_between_pages_and_edit_widgets(app: Page):
+    """Test that we can switch between pages and page widgets do not persist."""
+    app.get_by_test_id("stSidebarNav").locator("a").nth(1).click()
+    wait_for_app_run(app, wait_delay=1000)
+
+    slider = app.locator('.stSlider [role="slider"]').nth(1)
+    slider.click()
+    slider.press("ArrowRight")
+    wait_for_app_run(app)
+    expect(app.get_by_test_id("stMarkdown").nth(1)).to_contain_text("x is 1")
+
+    app.get_by_test_id("stSidebarNav").locator("a").nth(0).click()
+    wait_for_app_run(app, wait_delay=1000)
+
+    app.get_by_test_id("stSidebarNav").locator("a").nth(1).click()
+    wait_for_app_run(app, wait_delay=1000)
+
+    expect(app.get_by_test_id("stMarkdown").nth(1)).to_contain_text("x is 0")
+
+
+def test_titles_are_set_correctly(app: Page):
+    """Test that page titles work as expected"""
+    nav = app.get_by_test_id("stSidebarNav")
+    expect(nav.locator("a").nth(0)).to_contain_text("page 2")
```

Comment:
nit: instead of doing `nav.locator("a").nth(0)` in every test, we have started to use a pattern like:
```python
def _get_page2_link(app: Page) -> Locator:
  return nav.locator("a").nth(0)

# etc.
```
(see [here](https://github.com/streamlit/streamlit/blob/develop/e2e_playwright/st_dataframe_selections_test.py#L66-L87))
and use these convenient functions in the tests. It makes them easier to read and if the app ever is modified, e.g. new elements are added, then you have to change only a single line.

Additionally, it might make sense to move the expects such as 
```python
    expect(nav.locator("a").nth(0)).to_contain_text("page 2")
    expect(nav.locator("a").nth(1)).to_contain_text("Different Title")
    expect(nav.locator("a").nth(2)).to_contain_text("page 5")
    expect(nav.locator("a").nth(3)).to_contain_text("page 9")
```
into a convenient function and call that from every test that checks for it instead of duplicating it.